src/net/p2p.rs

```diff
@@ -3074,8 +3082,17 @@ impl PeerNetwork {
                             if always_allowed.len() == 0 {
                                 have_always_allowed = true;
                             } else {
+                                let mut have_unsynced = false;
                                 if let Some(ref inv_state) = self.inv_state {
                                     for (nk, stats) in inv_state.block_stats.iter() {
+                                        if self.is_bound(&nk) {
+                                            continue;
+                                        }
+                                        if Some((nk.addrbytes.clone(), nk.port))
+                                            == self.local_peer.public_ip_address
+                                        {
+                                            continue;
+                                        }
```

Comment:
Added clarification.  Basically, depending on your network topology, it can be the case that all always-allowed nodes are unroutable to you (i.e. they have the same same public IP address or something).  In that case, we need to declare the inv sync completed; otherwise it will stall forever.  That's what this extra flag does.  I added comments.